In [1]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier

from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

In [3]:
df=pd.read_csv('/content/fraudTest.csv')
df.head()

Unnamed: 0 trans_date_trans_time            cc_num  \
0           0   2020-06-21 12:14:25  2291163933867244   
1           1   2020-06-21 12:14:33  3573030041201292   
2           2   2020-06-21 12:14:53  3598215285024754   
3           3   2020-06-21 12:15:15  3591919803438423   
4           4   2020-06-21 12:15:17  3526826139003047   

                               merchant        category    amt   first  \
0                 fraud_Kirlin and Sons   personal_care   2.86    Jeff   
1                  fraud_Sporer-Keebler   personal_care  29.84  Joanne   
2  fraud_Swaniawski, Nitzsche and Welch  health_fitness  41.28  Ashley   
3                     fraud_Haley Group        misc_pos  60.05   Brian   
4                 fraud_Johnston-Casper          travel   3.19  Nathan   

       last gender                       street  ...      lat      long  \
0   Elliott      M            351 Darlene Green  ...  33.9659  -80.9355   
1  Williams      F             3638 Marsh Union  ...  40.3207 -110.4360   
2     Lopez      F         9333 Valentine Point  ...  40.6729  -73.5365   
3  Williams      M  32941 Krystal Mill Apt. 552  ...  28.5697  -80.8191   
4    Massey      M     5783 Evan Roads Apt. 465  ...  44.2529  -85.0170   

   city_pop                     job         dob  \
0    333497     Mechanical engineer  1968-03-19   
1       302  Sales professional, IT  1990-01-17   
2     34496       Librarian, public  1970-10-21   
3     54767            Set designer  1987-07-25   
4      1126      Furniture designer  1955-07-06   

                          trans_num     unix_time  merch_lat  merch_long  \
0  2da90c7d74bd46a0caf3777415b3ebd3  1.371817e+09  33.986391  -81.200714   
1  324cc204407e99f51b0d6ca0055005e7  1.371817e+09  39.450498 -109.960431   
2  c81755dbbbea9d5c77f094348a7579be  1.371817e+09  40.495810  -74.196111   
3  2159175b9efe66dc301f149d3d5abf8c  1.371817e+09  28.812398  -80.883061   
4  57ff021bd3f328f8738bb535c302a31b  1.371817e+09  44.959148  -85.884734   

   is_fraud  
0       0.0  
1       0.0  
2       0.0  
3       0.0  
4       0.0  

[5 rows x 23 columns]

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 58343 entries, 0 to 58342
Data columns (total 23 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Unnamed: 0             58343 non-null  int64  
 1   trans_date_trans_time  58343 non-null  object 
 2   cc_num                 58343 non-null  int64  
 3   merchant               58343 non-null  object 
 4   category               58343 non-null  object 
 5   amt                    58343 non-null  float64
 6   first                  58343 non-null  object 
 7   last                   58343 non-null  object 
 8   gender                 58343 non-null  object 
 9   street                 58343 non-null  object 
 10  city                   58343 non-null  object 
 11  state                  58343 non-null  object 
 12  zip                    58343 non-null  int64  
 13  lat                    58343 non-null  float64
 14  long                   58343 non-null  float64
 15  ci

In [5]:
df.describe()

Unnamed: 0        cc_num           amt           zip           lat  \
count  58343.000000  5.834300e+04  58343.000000  58343.000000  58343.000000   
mean   29171.000000  4.145401e+17     69.266936  48716.768370     38.543225   
std    16842.317715  1.305182e+18    146.402455  26820.868481      5.080845   
min        0.000000  6.041621e+10      1.000000   1257.000000     20.027100   
25%    14585.500000  1.800365e+14      9.700000  26237.000000     34.668900   
50%    29171.000000  3.518759e+15     47.130000  48174.000000     39.371600   
75%    43756.500000  4.633065e+15     82.835000  72011.000000     42.015800   
max    58342.000000  4.992346e+18  13149.150000  99783.000000     65.689900   

               long      city_pop     unix_time     merch_lat    merch_long  \
count  58343.000000  5.834300e+04  5.834200e+04  58342.000000  58342.000000   
mean     -90.148952  9.022664e+04  1.372663e+09     38.540999    -90.147824   
std       13.664820  3.044689e+05  5.092432e+05      5.115351     13.677735   
min     -165.672300  2.300000e+01  1.371817e+09     19.042320   -166.646251   
25%      -96.786900  7.430000e+02  1.372235e+09     34.748145    -96.836734   
50%      -87.456900  2.435000e+03  1.372620e+09     39.376865    -87.384904   
75%      -80.158000  2.032800e+04  1.373114e+09     41.978499    -80.224724   
max      -67.950300  2.906700e+06  1.373574e+09     66.669356    -66.955602   

           is_fraud  
count  58342.000000  
mean       0.004011  
std        0.063205  
min        0.000000  
25%        0.000000  
50%        0.000000  
75%        0.000000  
max        1.000000

In [6]:
df.isnull().sum()

Unnamed: 0               0
trans_date_trans_time    0
cc_num                   0
merchant                 0
category                 0
amt                      0
first                    0
last                     0
gender                   0
street                   0
city                     0
state                    0
zip                      0
lat                      0
long                     0
city_pop                 0
job                      0
dob                      0
trans_num                0
unix_time                1
merch_lat                1
merch_long               1
is_fraud                 1
dtype: int64

**Dropping Irrelevant Columns**

In [7]:
drop_cols = [
    'Unnamed: 0', 'cc_num', 'first', 'last',
    'street', 'trans_num', 'trans_date_trans_time', 'dob'
]
df = df.drop(columns=drop_cols)


**Handling Categorical Variables**

In [8]:
df = pd.get_dummies(df, drop_first=True)

**Split Features & Target**

In [9]:
X = df.drop('is_fraud', axis=1)
y = df['is_fraud']


**Train-Test Split**

In [11]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

**Feature Scaling**

In [12]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)


In [19]:
y.isna().sum()

np.int64(1)

In [20]:
X.isna().sum().sort_values(ascending=False).head()

merch_long     1
merch_lat      1
unix_time      1
city_Rumely    0
city_Sachse    0
dtype: int64

In [26]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)


In [29]:
df = df.dropna(subset=['is_fraud'])


In [30]:
X = df.drop('is_fraud', axis=1)
y = df['is_fraud']


In [31]:
from sklearn.impute import SimpleImputer

imputer = SimpleImputer(strategy='mean')
X = imputer.fit_transform(X)


In [34]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.3, random_state=42, stratify=y
)


In [35]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)


**Model Training & Evaluation**

In [36]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report

lr = LogisticRegression(max_iter=1000)
lr.fit(X_train, y_train)

y_pred_lr = lr.predict(X_test)

print("Logistic Regression Accuracy:", accuracy_score(y_test, y_pred_lr))
print(classification_report(y_test, y_pred_lr))


Logistic Regression Accuracy: 0.9953150888419128
              precision    recall  f1-score   support

         0.0       1.00      1.00      1.00     17433
         1.0       0.40      0.34      0.37        70

    accuracy                           1.00     17503
   macro avg       0.70      0.67      0.68     17503
weighted avg       0.99      1.00      1.00     17503



Class-wise Performance

Legitimate transactions (Class 0):

Precision: 1.00

Recall: 1.00

The model performs extremely well in identifying genuine transactions.

Fraudulent transactions (Class 1):

Precision: 0.40

Recall: 0.34

The model detects only 34% of actual fraud cases, meaning many fraudulent transactions are still missed.
